File: src/2020/day03/typescript/solution.ts
function formatInput(input: string): string[][] {
  return input
    .trim()
    .split('\n')
    .filter((l) => l.trim())
    .map((line) => line.split(''));
}

type Coordinate = [number, number];

function* iterateOnSlope(dr: number, dc: number): IterableIterator<Coordinate> {
  let currentY = 0,
    currentX = 0;

  while (true) {
    currentY += dr;
    currentX += dc;
    yield [currentY, currentX];
  }
}

function getTreeCount(area: string[][], rise: number, run: number): number {
  if (!area || !area.length || !area[0].length) {
    return 0;
  }
  const rowsN = area.length;
  const colsN = area[0].length;

  let treeCount = 0;

  const generator = iterateOnSlope(rise, run);
  let coords = generator.next();

  while (!coords.done) {
    let [y, x] = !coords.done ? coords.value! : [null, null];

    if (y == null || x == null) {
      break;
    }

    x %= colsN;

    if (y >= rowsN) {
      break;
    }

    treeCount += area[y][x] === '#' ? 1 : 0;

    coords = generator.next();
  }

  return treeCount;
}

export function part1(input: string): number | string | undefined {
  const area = formatInput(input);
  return getTreeCount(area, 1, 3);
}

export function part2(input: string): number | string | undefined {
  const area = formatInput(input);
  const slopes = [
    [1, 1],
    [1, 3],
    [1, 5],
    [1, 7],
    [2, 1],
  ];

  return slopes
    .map(([y, x]) => getTreeCount(area, y, x))
    .reduce((prod, num) => prod * num);
}

File: src/2020/day03/python/solution.py
import math
from typing import Generator


def format_list(input: str) -> list[tuple[str, ...]]:
    return [tuple(line) for line in input.strip().splitlines()]


def iterate_on_slope(
    dr: int, dc: int
) -> Generator[tuple[int, int]]:  # (current_r, current_c)
    common_divisor = math.gcd(dr, dc)

    if common_divisor == 0:
        return

    dr_unit = dr // common_divisor
    dc_unit = dc // common_divisor

    current_r, current_c = 0, 0

    yield (current_r, current_c)
    while True:
        current_r += dr_unit
        current_c += dc_unit
        yield (current_r, current_c)


def get_tree_count(area: list[tuple[str, ...]], rise: int, run: int) -> int:
    N_rows = len(area)
    N_cols = len(area[0])
    iterator = iterate_on_slope(rise, run)
    tree_count = 0

    while (coords := next(iterator, None)) is not None and coords[0] < N_rows:
        y, x = coords
        x %= N_cols

        if area[y][x] == "#":
            tree_count += 1

    return tree_count


def part1(input: str) -> int:
    area = format_list(input)
    return get_tree_count(area, 1, 3)


def part2(input: str) -> int:
    area = format_list(input)
    slopes = [(1, 1), (1, 3), (1, 5), (1, 7), (2, 1)]

    return math.prod(get_tree_count(area, rise, run) for rise, run in slopes)
